<img src="global_electronics_logo.jpeg" alt="Global Electronics Logo" width="400" height="50">

## Company Overview
Global Electronics Retailer is a multinational corporation specializing in the sale of electronic products and accessories. The company operates numerous retail stores across various countries, catering to a diverse customer base with a wide range of products. These products include various brands and categories, all meticulously tracked and managed to ensure optimal inventory control and sales performance.

## Project Objective
The primary objective of this project is to analyze the data generated by Global Electronics Retailer's operations to gain insights into sales performance, customer behavior, product popularity, and store efficiency. This analysis will help the company to make data-driven decisions aimed at enhancing customer satisfaction, optimizing inventory management, and improving overall business performance.

## Entity Relationship Diagram

<img src="global_electronics_erd.jpeg" alt="Global Electronics ERD" width="1000" height="50">

### STEP 1: CREATING THE SCHEMA

In [11]:
%%sql
-- CREATE customers TABLE
CREATE TABLE customers (
    customer_key INT,
    gender VARCHAR(10) CHECK (gender IN ('Male', 'Female')),
    name VARCHAR(50),
    city VARCHAR(50),
    state_code VARCHAR(50),
    state VARCHAR(50),
    zip_code VARCHAR(50),
    country VARCHAR(50),
    continent VARCHAR(50),
    birthday VARCHAR(10),
PRIMARY KEY (customer_key)
);

-- CREATE products TABLE
CREATE TABLE products (
    product_key INT,
    product_name VARCHAR(100),
    brand VARCHAR(50),
    color VARCHAR(50),
    unit_cost_usd VARCHAR(10),
    unit_price_usd VARCHAR(10),
    subcategory_key INT,
    subcategory VARCHAR(50),
    category_key INT,
    category VARCHAR(50),
PRIMARY KEY (product_key)
);

-- CREATE stores TABLE
CREATE TABLE stores (
    store_key INT,
    country VARCHAR(50),
    state VARCHAR(50),
    square_meters INT,
    open_date VARCHAR(50),
PRIMARY KEY (store_key)
);

-- CREATE exchange_rates TABLE
CREATE TABLE exchange_rates (
    date DATE,
    currency CHAR(3),
    exchange REAL
);

-- CREATE sales TABLE
CREATE TABLE sales (
    order_number INT,
    line_item INT,
    order_date VARCHAR(10),
    delivery_date VARCHAR(10),
    customer_key INT,
    store_key INT,
    product_key INT,
    quantity INT,
    currency_code CHAR(3),
FOREIGN KEY (customer_key) REFERENCES customers(customer_key),
FOREIGN KEY (store_key) REFERENCES stores(store_key),
FOREIGN KEY (product_key) REFERENCES products(product_key)
);

 * postgresql://postgres:***@localhost:5432/global_electronics
(psycopg2.errors.DuplicateTable) relation "customers" already exists

[SQL: -- CREATE customers TABLE
CREATE TABLE customers (
    customer_key INT,
    gender VARCHAR(10),
    name VARCHAR(50),
    city VARCHAR(50),
    state_code CHAR(3),
    state VARCHAR(50),
    zip_code INT,
    country VARCHAR(50),
    continent VARCHAR(50),
    birthday DATE,
PRIMARY KEY (customer_key)
);]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [100]:
%%sql
SELECT table_name
FROM information_schema.tables
WHERE table_schema != 'pg_catalog' AND table_schema != 'information_schema'

 * postgresql://postgres:***@localhost:5432/global_electronics
5 rows affected.


[('products',), ('customers',), ('stores',), ('exchange_rates',), ('sales',)]

### STEP 2: IMPORTING THE CSV FILES INTO THE SCHEMA

In [53]:
%%sql
-- import customers csv into customers table schema
COPY customers (customer_key, gender, name, city, state_code, state, zip_code, country, continent, birthday)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\global_electronic_data\Customers2.csv'
DELIMITER ','
CSV HEADER;

-- import products csv into product table schema
%%sql
COPY products (product_key, product_name, brand, color, unit_cost_usd, unit_price_usd, subcategory_key, subcategory, category_key, category)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\global_electronic_data\Products.csv'
DELIMITER ','
CSV HEADER;

-- import stores csv into stores table schema
%%sql
COPY stores (store_key, country, state, square_meters, open_date)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\global_electronic_data\Stores.csv'
DELIMITER ','
CSV HEADER;

-- import exchange_rates csv into exchange_rates table schema   
%%sql
COPY exchange_rates (date, currency, exchange)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\global_electronic_data\Exchange_Rates.csv'
DELIMITER ','
CSV HEADER;

-- import sales csv into sales table schema  
%%sql
COPY sales (order_number, line_item, order_date, delivery_date, customer_key, store_key, product_key, quantity, currency_code)
FROM 'C:\Program Files\PostgreSQL\16\scripts\Dataset\global_electronic_data\Sales.csv'
DELIMITER ','
CSV HEADER;

 * postgresql://postgres:***@localhost:5432/global_electronics
15266 rows affected.


[]

## BUSINESS OBJECTIVES AND QUESTIONS

## Identifying regional product preferences
- Find the top 3 most popular product subcategories for customers located in each continent, based on the total quantity sold, and display the subcategory name, category, and the corresponding total quantity sold for each continent.

In [182]:
%%sql
WITH temp_table AS (
    SELECT c.continent,
        p.subcategory,
        p.category,
        SUM(s.quantity) AS total_quantity
    FROM sales AS s 
    LEFT JOIN customers AS c
        ON s.customer_key = c.customer_key
    LEFT JOIN products AS p
        ON s.product_key = p.product_key
    GROUP BY c.continent, p.subcategory, p.category
),
ranked_subcategories AS (
SELECT continent,
    subcategory,
    category,
    total_quantity,
    RANK() OVER(PARTITION BY continent ORDER BY total_quantity DESC) AS rank
FROM temp_table
)
SELECT *
FROM ranked_subcategories
WHERE rank IN (1, 2, 3)

 * postgresql://postgres:***@localhost:5432/global_electronics
9 rows affected.


[('Australia', 'Movie DVD', 'Music, Movies and Audio Books', 1388, 1),
 ('Australia', 'Desktops', 'Computers', 835, 2),
 ('Australia', 'Bluetooth Headphones', 'Audio', 676, 3),
 ('Europe', 'Movie DVD', 'Music, Movies and Audio Books', 9629, 1),
 ('Europe', 'Desktops', 'Computers', 6940, 2),
 ('Europe', 'Bluetooth Headphones', 'Audio', 4486, 3),
 ('North America', 'Movie DVD', 'Music, Movies and Audio Books', 17785, 1),
 ('North America', 'Desktops', 'Computers', 12851, 2),
 ('North America', 'Bluetooth Headphones', 'Audio', 8491, 3)]

## Evaluating top-performing store-category combinations
- Identify the stores that have consistently generated the highest revenue for a specific product category during the last three years, considering the exchange rates at the time of each sale.

In [46]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
    ),
cleaned_sales AS (
    SELECT order_number, line_item, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, customer_key, store_key, product_key, quantity, currency_code
    FROM sales
    ),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
),
store_performance AS (
    SELECT cs.store_key, cp.category,
           SUM(CASE WHEN cs.currency_code IN ('USD', 'EUR', 'AUD', 'CAD', 'GBP') THEN (cs.quantity*cp.unit_price_usd*ce.exchange)
                ELSE 000 END) AS revenue
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        ON cs.product_key = cp.product_key
    WHERE cs.order_date >= DATE_TRUNC('month', (SELECT MAX(order_date) FROM cleaned_sales)) - INTERVAL '3 year'
    GROUP BY cs.store_key, cp.category
),
store_rank AS (
    SELECT state, country, category, revenue,
        RANK() OVER(PARTITION BY category ORDER BY revenue DESC) AS rank
    FROM store_performance AS sp
    INNER JOIN stores AS s
        ON sp.store_key = s.store_key
)
SELECT state, country, category, ROUND(revenue::NUMERIC, 2) AS revenue
FROM store_rank
WHERE rank = 2

 * postgresql://postgres:***@localhost:5432/global_electronics
8 rows affected.


[('Northwest Territories', 'Canada', 'Audio', Decimal('70167.98')),
 ('Nevada', 'United States', 'Cameras and camcorders', Decimal('150194.41')),
 ('Northwest Territories', 'Canada', 'Cell phones', Decimal('139748.86')),
 ('Northwest Territories', 'Canada', 'Computers', Decimal('523061.40')),
 ('Northwest Territories', 'Canada', 'Games and Toys', Decimal('19694.06')),
 ('Kansas', 'United States', 'Home Appliances', Decimal('199430.00')),
 ('Nunavut', 'Canada', 'Music, Movies and Audio Books', Decimal('71480.59')),
 ('Alaska', 'United States', 'TV and Video', Decimal('140309.20'))]

INSIGHTS 💡💡 <br>
• The highest revenue was generated through the online channel across all product categories. However, the physical stores listed above achieved the highest revenue for each individual product category. <br>
• Apart from the online channel, 62.5% of the top revenue-generating stores for each product category were located in the Canada and the remaining 37.5% in the United States, highlighting strong profitability in these markets.

## Analyzing high-value customer segments
- Calculate the average customer lifetime value (total revenue generated) for each combination of gender, age group (based on birthday), and country, and display the top 5 combinations with the highest lifetime value.

In [12]:
%%sql
-- CATEGORIZING CUSTOMERS BY AGE GROUP
WITH bins AS (
    SELECT generate_series(10, 80, 10) AS lower,
           generate_series(19, 90, 10) AS upper
),
customer_age AS (
    SELECT DISTINCT s.customer_key,
           c.gender,
           c.country,
           (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) AS age
    FROM sales AS s
    INNER JOIN customers AS c
        USING(customer_key)
)
SELECT
    lower AS lower_age_range,
    upper AS upper_age_range,
    COUNT(age) AS no_of_customers
FROM bins AS b
LEFT JOIN customer_age AS c
    ON age >= lower AND age <= upper
GROUP BY lower, upper
ORDER BY lower

 * postgresql://postgres:***@localhost:5432/global_electronics
8 rows affected.


[(10, 19, 13),
 (20, 29, 1808),
 (30, 39, 1807),
 (40, 49, 1732),
 (50, 59, 1790),
 (60, 69, 1791),
 (70, 79, 1751),
 (80, 89, 1195)]

INSIGHTS 💡💡 <br>
• The relatively stable number of customers from 20 to 79 years old suggests strong and sustained engagement across a wide age range, highlighting the company’s broad appeal.<br>
• The company maintains a significant number of older customers (70-89 years), indicating the importance of considering the needs and preferences of aging customers in product development and marketing strategies.

In [44]:
%%sql
WITH customer_temp AS (
    SELECT customer_key,
    gender,
    country,
    (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) AS age,
    CASE WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 10 AND 19 THEN '<20'
         WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 20 AND 29 THEN '20-29'
         WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 30 AND 39 THEN '30-39'
         WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 40 AND 49 THEN '40-49'
         WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 50 AND 59 THEN '50-59'
         WHEN (SELECT MAX(EXTRACT(YEAR FROM (TO_DATE(order_date, 'MM-DD-YYYY')))) FROM sales) - EXTRACT(YEAR FROM birthday) BETWEEN 60 AND 90 THEN '60 and above'
         ELSE 'Outlier' END AS age_group
    FROM customers
),
cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
),
cleaned_sales AS (
    SELECT order_number, line_item, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, customer_key, store_key, product_key, quantity, currency_code
    FROM sales
),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
),
customer_group AS (
    SELECT ct.gender, ct.country, ct.age_group,
           SUM(CASE WHEN cs.currency_code = 'USD' THEN (cs.quantity*cp.unit_price_usd)
                ELSE (cs.quantity*cp.unit_price_usd)/ce.exchange
                END) AS revenue
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        ON cs.product_key = cp.product_key
    LEFT JOIN customer_temp AS ct
        ON cs.customer_key = ct.customer_key
    GROUP BY ct.gender, ct.country, ct.age_group
)
SELECT gender, country, age_group, ROUND(revenue::NUMERIC, 2) AS revenue
FROM customer_group
ORDER BY revenue DESC
LIMIT 5

 * postgresql://postgres:***@localhost:5432/global_electronics
5 rows affected.


[('Male', 'United States', '60 and above', Decimal('6560789.70')),
 ('Female', 'United States', '60 and above', Decimal('5784863.97')),
 ('Female', 'United States', '40-49', Decimal('2416179.58')),
 ('Female', 'United States', '20-29', Decimal('2258987.98')),
 ('Male', 'United States', '50-59', Decimal('2226946.34'))]

INSIGHTS 💡💡<br>
• The presence of both male and female customers in the '60 and above' age range among the top categories highlights the significant revenue potential from elderly customers.<br>
• Female customers in their 20s, 40s and 60s all feature in the top revenue categories, suggesting that the company’s products or services appeal to a wide range of life stages for women.

## Monitoring store-level revenue growth trend
- Calculate the year-over-year growth rate in revenue for each store, considering the exchange rates at the time of each sale, and identify the stores that have consistently shown positive growth over the last 3 years.

In [90]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
    ),
cleaned_sales AS (
    SELECT order_number, line_item, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, customer_key, store_key, product_key, quantity, currency_code
    FROM sales
    ),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
    ),
store_performance AS (
    SELECT cs.store_key, cs.order_date,
           SUM(CASE WHEN cs.currency_code IN ('USD', 'EUR', 'AUD', 'CAD', 'GBP') THEN (cs.quantity*cp.unit_price_usd*ce.exchange)
                ELSE 000 END) AS revenue
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        ON cs.product_key = cp.product_key
    WHERE cs.order_date < DATE_TRUNC('year', (SELECT MAX(order_date) FROM cleaned_sales))
    GROUP BY cs.store_key, cs.order_date
    ),
store_yearly_rev AS (
    SELECT state AS store, country, EXTRACT(YEAR FROM order_date) AS year, ROUND(SUM(revenue::NUMERIC), 2) AS revenue
    FROM store_performance AS sr
    INNER JOIN stores AS s
        ON sr.store_key = s.store_key
    GROUP BY store, country, year
    ORDER BY store, year
    ),
store_rev_history AS (
    SELECT *, LAG(revenue) OVER(PARTITION BY store ORDER by store, year) AS prev_year_rev
    FROM store_yearly_rev
    ),
percent_growth AS (
    SELECT store, country, year, '$'|| ROUND(revenue, 0) AS revenue, '$'|| ROUND(prev_year_rev, 0) AS prev_year_rev, ROUND(((revenue-prev_year_rev)/prev_year_rev)*100, 2) ||'%' AS percent_growth
    FROM store_rev_history
    WHERE revenue > prev_year_rev 
    ),
final_grouping AS (
    SELECT *
    FROM percent_growth
    WHERE store IN (
        SELECT store
        FROM percent_growth
        GROUP BY store)
)
SELECT store, country, COUNT(*) AS positive_growth_years
FROM final_grouping
GROUP BY store, country
HAVING COUNT(*) >= 3
ORDER BY store

 * postgresql://postgres:***@localhost:5432/global_electronics
25 rows affected.


[('Alaska', 'United States', 3),
 ('Arkansas', 'United States', 3),
 ('Armagh', 'United Kingdom', 3),
 ('Australian Capital Territory', 'Australia', 3),
 ('Basse-Normandie', 'France', 3),
 ('Blaenau Gwent', 'United Kingdom', 3),
 ('Connecticut', 'United States', 3),
 ('Drenthe', 'Netherlands', 3),
 ('Dungannon and South Tyrone', 'United Kingdom', 3),
 ('Freistaat Thüringen', 'Germany', 3),
 ('Hamburg', 'Germany', 3),
 ('Kansas', 'United States', 3),
 ('Montana', 'United States', 3),
 ('Nebraska', 'United States', 3),
 ('New Hampshire', 'United States', 3),
 ('New Mexico', 'United States', 3),
 ('Newfoundland and Labrador', 'Canada', 3),
 ('North Down', 'United Kingdom', 3),
 ('Nunavut', 'Canada', 3),
 ('Online', 'Online', 3),
 ('Oregon', 'United States', 3),
 ('Pesaro', 'Italy', 3),
 ('Tasmania', 'Australia', 3),
 ('West Virginia', 'United States', 3),
 ('Western Australia', 'Australia', 3)]

In [84]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
    ),
cleaned_sales AS (
    SELECT order_number, line_item, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, customer_key, store_key, product_key, quantity, currency_code
    FROM sales
    ),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
    ),
store_performance AS (
    SELECT cs.store_key, cs.order_date,
           SUM(CASE WHEN cs.currency_code IN ('USD', 'EUR', 'AUD', 'CAD', 'GBP') THEN (cs.quantity*cp.unit_price_usd*ce.exchange)
                ELSE 000 END) AS revenue
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        ON cs.product_key = cp.product_key
    WHERE cs.order_date < DATE_TRUNC('year', (SELECT MAX(order_date) FROM cleaned_sales))
    GROUP BY cs.store_key, cs.order_date
    ),
store_yearly_rev AS (
    SELECT state AS store, country, EXTRACT(YEAR FROM order_date) AS year, ROUND(SUM(revenue::NUMERIC), 2) AS revenue
    FROM store_performance AS sr
    INNER JOIN stores AS s
        ON sr.store_key = s.store_key
    GROUP BY store, country, year
    ORDER BY store, year
    ),
store_rev_history AS (
    SELECT *, LAG(revenue) OVER(PARTITION BY store ORDER by store, year) AS prev_year_rev
    FROM store_yearly_rev
    ),
percent_growth AS (
    SELECT store, country, year, '$'|| ROUND(revenue, 0) AS revenue, '$'|| ROUND(prev_year_rev, 0) AS prev_year_rev, ROUND(((revenue-prev_year_rev)/prev_year_rev)*100, 2) ||'%' AS percent_growth
    FROM store_rev_history
    WHERE revenue > prev_year_rev 
    ) 
SELECT *
    FROM percent_growth
    WHERE store IN (
        SELECT store
        FROM percent_growth
        GROUP BY store
)

 * postgresql://postgres:***@localhost:5432/global_electronics
128 rows affected.


[('Alaska', 'United States', Decimal('2017'), '$179887', '$116426', '54.51%'),
 ('Alaska', 'United States', Decimal('2018'), '$190944', '$179887', '6.15%'),
 ('Alaska', 'United States', Decimal('2019'), '$469094', '$190944', '145.67%'),
 ('Arkansas', 'United States', Decimal('2017'), '$186022', '$138512', '34.30%'),
 ('Arkansas', 'United States', Decimal('2018'), '$292707', '$186022', '57.35%'),
 ('Arkansas', 'United States', Decimal('2019'), '$374484', '$292707', '27.94%'),
 ('Armagh', 'United Kingdom', Decimal('2017'), '$93446', '$74089', '26.13%'),
 ('Armagh', 'United Kingdom', Decimal('2018'), '$138632', '$93446', '48.35%'),
 ('Armagh', 'United Kingdom', Decimal('2019'), '$224661', '$138632', '62.06%'),
 ('Australian Capital Territory', 'Australia', Decimal('2017'), '$32444', '$22139', '46.54%'),
 ('Australian Capital Territory', 'Australia', Decimal('2018'), '$74408', '$32444', '129.35%'),
 ('Australian Capital Territory', 'Australia', Decimal('2019'), '$120501', '$74408', '61.95%'),
 ('Ayrshire', 'United Kingdom', Decimal('2018'), '$165436', '$96627', '71.21%'),
 ('Basse-Normandie', 'France', Decimal('2017'), '$14234', '$13050', '9.07%'),
 ('Basse-Normandie', 'France', Decimal('2018'), '$34758', '$14234', '144.20%'),
 ('Basse-Normandie', 'France', Decimal('2019'), '$47367', '$34758', '36.28%'),
 ('Belfast', 'United Kingdom', Decimal('2018'), '$183812', '$112587', '63.26%'),
 ('Belfast', 'United Kingdom', Decimal('2019'), '$223591', '$183812', '21.64%'),
 ('Berlin', 'Germany', Decimal('2018'), '$133244', '$60446', '120.43%'),
 ('Berlin', 'Germany', Decimal('2019'), '$145630', '$133244', '9.30%'),
 ('Blaenau Gwent', 'United Kingdom', Decimal('2017'), '$110430', '$55909', '97.52%'),
 ('Blaenau Gwent', 'United Kingdom', Decimal('2018'), '$134245', '$110430', '21.57%'),
 ('Blaenau Gwent', 'United Kingdom', Decimal('2019'), '$264929', '$134245', '97.35%'),
 ('Brandenburg', 'Germany', Decimal('2017'), '$69022', '$46102', '49.72%'),
 ('Brandenburg', 'Germany', Decimal('2018'), '$145483', '$69022', '110.78%'),
 ('Connecticut', 'United States', Decimal('2017'), '$187282', '$123419', '51.74%'),
 ('Connecticut', 'United States', Decimal('2018'), '$294591', '$187282', '57.30%'),
 ('Connecticut', 'United States', Decimal('2019'), '$407443', '$294591', '38.31%'),
 ('Corse', 'France', Decimal('2017'), '$26468', '$13851', '91.10%'),
 ('Corse', 'France', Decimal('2019'), '$41471', '$15492', '167.69%'),
 ('Drenthe', 'Netherlands', Decimal('2017'), '$38862', '$35860', '8.37%'),
 ('Drenthe', 'Netherlands', Decimal('2018'), '$92236', '$38862', '137.34%'),
 ('Drenthe', 'Netherlands', Decimal('2019'), '$97432', '$92236', '5.63%'),
 ('Dungannon and South Tyrone', 'United Kingdom', Decimal('2017'), '$119546', '$83283', '43.54%'),
 ('Dungannon and South Tyrone', 'United Kingdom', Decimal('2018'), '$162744', '$119546', '36.13%'),
 ('Dungannon and South Tyrone', 'United Kingdom', Decimal('2019'), '$228032', '$162744', '40.12%'),
 ('Enna', 'Italy', Decimal('2018'), '$187460', '$108149', '73.33%'),
 ('Enna', 'Italy', Decimal('2019'), '$248178', '$187460', '32.39%'),
 ('Fermanagh', 'United Kingdom', Decimal('2017'), '$125014', '$82518', '51.50%'),
 ('Flevoland', 'Netherlands', Decimal('2018'), '$76290', '$15195', '402.07%'),
 ('Flevoland', 'Netherlands', Decimal('2019'), '$129391', '$76290', '69.60%'),
 ('Franche-Comté', 'France', Decimal('2019'), '$23043', '$12773', '80.41%'),
 ('Freie Hansestadt Bremen', 'Germany', Decimal('2019'), '$210430', '$71265', '195.28%'),
 ('Freistaat Thüringen', 'Germany', Decimal('2017'), '$70360', '$53839', '30.69%'),
 ('Freistaat Thüringen', 'Germany', Decimal('2018'), '$132181', '$70360', '87.86%'),
 ('Freistaat Thüringen', 'Germany', Decimal('2019'), '$242590', '$132181', '83.53%'),
 ('Friesland', 'Netherlands', Decimal('2019'), '$118848', '$46710', '154.44%'),
 ('Groningen', 'Netherlands', Decimal('2017'), '$51403', '$39983', '28.56%'),
 ('Groningen', 'Netherlands', Decimal('2018'), '$120098', '$51403', '133.64%'),
 (

INSIGHTS 💡💡<br>
• 25 out of the 67 stores (i.e 37.3%) had year-on-year growth between 2016 and 2019.<br>
• 64% of these growing stores are located in the United States.<br>
• However, it was observed that all the stores recorded a drop in revenue in 2020 when compared to 2019. This can be attributed to the effect of the pandemic and various physical restrictions placed on customers.<br>

## Determining localized subcategory profitability
- Find the product subcategories that have generated the highest revenue for each state, and display the subcategory name, category, state, and the corresponding total revenue.

In [47]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
    ),
cleaned_sales AS (
    SELECT order_number, line_item, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, customer_key, store_key, product_key, quantity, currency_code
    FROM sales
    ),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
),
store_performance AS (
    SELECT cs.store_key, cp.subcategory, cp.category,
           SUM(CASE WHEN cs.currency_code IN ('USD', 'EUR', 'AUD', 'CAD', 'GBP') THEN (cs.quantity*cp.unit_price_usd*ce.exchange)
                ELSE 000 END) AS revenue
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        ON cs.product_key = cp.product_key
    GROUP BY cs.store_key, cp.subcategory, cp.category
),
store_best_prod AS (
    SELECT state, subcategory, category, revenue,
        RANK() OVER(PARTITION BY state ORDER BY revenue DESC) AS rank
    FROM store_performance AS sp
    INNER JOIN stores AS s
        ON sp.store_key = s.store_key
)
SELECT state, subcategory, category, ROUND(revenue::NUMERIC, 0) AS revenue
FROM store_best_prod
WHERE rank = 1
ORDER BY state

 * postgresql://postgres:***@localhost:5432/global_electronics
58 rows affected.


[('Alaska', 'Desktops', 'Computers', Decimal('194475')),
 ('Arkansas', 'Desktops', 'Computers', Decimal('208479')),
 ('Armagh', 'Desktops', 'Computers', Decimal('137197')),
 ('Australian Capital Territory', 'Desktops', 'Computers', Decimal('52055')),
 ('Ayrshire', 'Desktops', 'Computers', Decimal('142737')),
 ('Basse-Normandie', 'Desktops', 'Computers', Decimal('35649')),
 ('Belfast', 'Desktops', 'Computers', Decimal('122019')),
 ('Berlin', 'Desktops', 'Computers', Decimal('100052')),
 ('Blaenau Gwent', 'Desktops', 'Computers', Decimal('101601')),
 ('Brandenburg', 'Desktops', 'Computers', Decimal('54964')),
 ('Caltanissetta', 'Water Heaters', 'Home Appliances', Decimal('19984')),
 ('Connecticut', 'Desktops', 'Computers', Decimal('258032')),
 ('Corse', 'Desktops', 'Computers', Decimal('22994')),
 ('Drenthe', 'Desktops', 'Computers', Decimal('74848')),
 ('Dungannon and South Tyrone', 'Desktops', 'Computers', Decimal('121763')),
 ('Enna', 'Desktops', 'Computers', Decimal('145251')),
 ('Fermanagh', 'Desktops', 'Computers', Decimal('49757')),
 ('Flevoland', 'Desktops', 'Computers', Decimal('59599')),
 ('Franche-Comté', 'Desktops', 'Computers', Decimal('27549')),
 ('Freie Hansestadt Bremen', 'Desktops', 'Computers', Decimal('76954')),
 ('Freistaat Thüringen', 'Desktops', 'Computers', Decimal('98553')),
 ('Friesland', 'Desktops', 'Computers', Decimal('75966')),
 ('Groningen', 'Desktops', 'Computers', Decimal('52155')),
 ('Hamburg', 'Desktops', 'Computers', Decimal('110591')),
 ('Hawaii', 'Desktops', 'Computers', Decimal('169562')),
 ('Hessen', 'Televisions', 'TV and Video', Decimal('77958')),
 ('Idaho', 'Desktops', 'Computers', Decimal('181518')),
 ('Iowa', 'Desktops', 'Computers', Decimal('173914')),
 ('Kansas', 'Desktops', 'Computers', Decimal('233679')),
 ('La Réunion', 'Desktops', 'Computers', Decimal('40648')),
 ('Limousin', 'Desktops', 'Computers', Decimal('19384')),
 ('Maine', 'Desktops', 'Computers', Decimal('252154')),
 ('Martinique', 'Camcorders', 'Cameras and camcorders', Decimal('17332')),
 ('Mayotte', 'Desktops', 'Computers', Decimal('36073')),
 ('Montana', 'Desktops', 'Computers', Decimal('216612')),
 ('Nebraska', 'Desktops', 'Computers', Decimal('269945')),
 ('Nevada', 'Desktops', 'Computers', Decimal('248245')),
 ('New Hampshire', 'Desktops', 'Computers', Decimal('241222')),
 ('New Mexico', 'Desktops', 'Computers', Decimal('197794')),
 ('Newfoundland and Labrador', 'Desktops', 'Computers', Decimal('280768')),
 ('North Down', 'Desktops', 'Computers', Decimal('153484')),
 ('Northern Territory', 'Televisions', 'TV and Video', Decimal('8019')),
 ('Northwest Territories', 'Desktops', 'Computers', Decimal('342241')),
 ('Nunavut', 'Desktops', 'Computers', Decimal('202542')),
 ('Online', 'Desktops', 'Computers', Decimal('2126798')),
 ('Oregon', 'Desktops', 'Computers', Decimal('206263')),
 ('Pesaro', 'Desktops', 'Computers', Decimal('155068')),
 ('Saarland', 'Desktops', 'Computers', Decimal('49792')),
 ('Sachsen-Anhalt', 'Desktops', 'Computers', Decimal('87432')),
 ('South Carolina', 'Desktops', 'Computers', Decimal('221574')),
 ('South Dakota', 'Desktops', 'Computers', Decimal('132036')),
 ('Tasmania', 'Televisions', 'TV and Video', Decimal('119997')),
 ('Utah', 'Desktops', 'Computers', Decimal('139304')),
 ('Victoria', 'Desktops', 'Computers', Decimal('161026')),
 ('Washington DC', 'Desktops', 'Computers', Decimal('225507')),
 ('West Virginia', 'Desktops', 'Computers', Decimal('238931')),
 ('Western Australia', 'Desktops', 'Computers', Decimal('111398')),
 ('Wyoming', 'Desktops', 'Computers', Decimal('203545'))]

INSIGHTS 💡💡<br>
• Desktops sold highest in over 91.4% of the stores. This is followed by Televisions (5%). Water Heaters and Camcorders each represent 1.7%.<br>
• This suggests that the company's most-demanded product are Desktops, but with such a high reliance on Desktops, there may be an opportunity for the company to diversify its product offerings to reduce dependence on a single product category.

## Identifying truly loyal customer base
- Identify the best 20 customers who have made purchases from at least 7 stores within a specific country, and display their name, country, and the total amount spent across all stores in that country.

In [98]:
%%sql
WITH customer_agg AS (
    SELECT customer_key, COUNT(distinct store_key) as stores_visited
    FROM sales 
    GROUP BY customer_key
),
cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
),
sales_agg_by_cust AS (
    SELECT c.customer_key, c.name, c.country, SUM(s.quantity*cp.unit_price_usd) AS total_amount_spent
    FROM sales AS s
    LEFT JOIN cleaned_product AS cp
        ON s.product_key = cp.product_key
    LEFT JOIN customers AS c
        ON s.customer_key = c.customer_key
    GROUP BY c.customer_key, c.name, c.country
)
SELECT customer_key, name, country, stores_visited, '$'|| ROUND(total_amount_spent::DECIMAL, 0) AS total_amount_spent
FROM customer_agg AS ca
INNER JOIN sales_agg_by_cust AS sabc
    USING(customer_key)
WHERE ca.stores_visited >= 7
ORDER BY ca.stores_visited DESC
LIMIT 20

 * postgresql://postgres:***@localhost:5432/global_electronics
20 rows affected.


[(1925694, 'Fabrice Lamoureux', 'United States', 8, '$26339'),
 (2034922, 'Aubrey Audet', 'United States', 8, '$23992'),
 (1737466, 'Christina Miller', 'United States', 8, '$4817'),
 (1224615, 'Paula Barker', 'United States', 8, '$12659'),
 (2080484, 'Delmer Martinez', 'United States', 8, '$19536'),
 (1328055, 'Robert Smith', 'United States', 8, '$24736'),
 (1610098, 'Sara Christensen', 'United States', 8, '$20458'),
 (1702221, 'Matthew Flemming', 'United States', 8, '$61872'),
 (2014226, 'Theodor Kristoffersen', 'United States', 8, '$16772'),
 (1523982, 'Robert Nelsen', 'United States', 8, '$29009'),
 (1388023, 'David Breton', 'United States', 8, '$25237'),
 (1349131, 'Katherine Rosales', 'United States', 8, '$16479'),
 (1397271, 'Yesenia Chivers', 'United States', 8, '$10502'),
 (1521733, 'Lisa Sherman', 'United States', 8, '$17474'),
 (1579183, 'Paul Warren', 'United States', 8, '$35780'),
 (1268661, 'Michael Whitehurst', 'United States', 8, '$20814'),
 (1312078, 'Joseph Garner', 'United States', 7, '$20079'),
 (1493457, 'Marcos McMichael', 'United States', 7, '$26102'),
 (1602426, 'Joseph Hamilton', 'United States', 7, '$14129'),
 (1388042, 'Kenneth Tucker', 'United States', 7, '$16882')]

INSIGHTS 💡💡<br>
• These 20 customers have visited at least 7 different stores to make an order. These customers can be said to be most loyal customers.

## Optimizing average order value drivers
- Calculate the average order value (total order amount divided by the number of items) and display the top 5 and bottom 5 stores with the highest and lowest average order values respectively.

In [48]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
),
cleaned_sales AS (
    SELECT order_number, customer_key, TO_DATE(order_date, 'MM-DD-YYYY') AS order_date, store_key, product_key, quantity, currency_code    
    FROM sales
),
cleaned_exchange_rates AS (
    SELECT TO_DATE(date, 'MM-DD-YYYY') AS date, currency, exchange
    FROM exchange_rates
),
orders AS (
    SELECT cs.order_number, s.state,
           SUM(CASE WHEN cs.currency_code IN ('USD', 'EUR', 'AUD', 'CAD', 'GBP') THEN (cs.quantity*cp.unit_price_usd*ce.exchange)
                ELSE 000 END) AS revenue,
            SUM(quantity) AS total_quantity
    FROM cleaned_sales AS cs
    INNER JOIN cleaned_exchange_rates AS ce
        ON cs.order_date = ce.date AND cs.currency_code = ce.currency
    LEFT JOIN cleaned_product AS cp
        USING(product_key)
    LEFT JOIN stores AS s
        USING(store_key) 
    GROUP BY cs.order_number, s.store_key
),
store_group AS (
    SELECT state AS store, SUM(revenue) AS store_total_revenue, SUM(total_quantity) AS store_total_quantity
    FROM orders AS og
    GROUP BY state
),
performance AS (
    SELECT store, '$'|| ROUND(store_total_revenue::DECIMAL/store_total_quantity, 0) AS avg_order_value
    FROM store_group
    ORDER BY avg_order_value DESC
),
best_performing AS (
    SELECT *
    FROM performance
    ORDER BY avg_order_value DESC
    LIMIT 5
),
least_performing AS (
    SELECT *
    FROM performance
    ORDER BY avg_order_value ASC
    LIMIT 5
)
SELECT *, 'Best-Performing' AS performance
FROM best_performing
UNION ALL
SELECT '--------', '------', '-----------------'
UNION ALL
SELECT *, 'Least-Performing'
FROM least_performing


 * postgresql://postgres:***@localhost:5432/global_electronics
11 rows affected.


[('Tasmania', '$478', 'Best-Performing'),
 ('Victoria', '$407', 'Best-Performing'),
 ('Australian Capital Territory', '$393', 'Best-Performing'),
 ('Western Australia', '$390', 'Best-Performing'),
 ('Newfoundland and Labrador', '$386', 'Best-Performing'),
 ('--------', '------', '-----------------'),
 ('Armagh', '$199', 'Least-Performing'),
 ('Blaenau Gwent', '$209', 'Least-Performing'),
 ('Fermanagh', '$212', 'Least-Performing'),
 ('Ayrshire', '$213', 'Least-Performing'),
 ('Basse-Normandie', '$214', 'Least-Performing')]

## Princing Statistics
- Find the total number of products with unit prices higher than the average for their respective categories.

In [130]:
%%sql
WITH cleaned_product AS (
    SELECT product_key, product_name, REPLACE(REPLACE(unit_cost_usd, ',', ''), '$', '')::NUMERIC AS unit_cost_usd,
            REPLACE(REPLACE(unit_price_usd, '$', ''), ',', '')::NUMERIC AS unit_price_usd,
            subcategory_key, subcategory, category_key, category
    FROM products
    ),
total_products AS (
    SELECT product_key, product_name, category, unit_price_usd
    FROM cleaned_product AS csp1
    WHERE unit_price_usd >
    (SELECT AVG(unit_price_usd) FROM cleaned_product AS csp2 WHERE csp1.category_key = csp2.category_key)
    ORDER BY unit_price_usd DESC
)
SELECT COUNT(*) AS higher_priced_products,
        (SELECT COUNT(*) FROM PRODUCTS) AS total_no_of_products,
        COUNT(*)*100/(SELECT COUNT(*) FROM PRODUCTS)||'%' AS percentage
FROM total_products

 * postgresql://postgres:***@localhost:5432/global_electronics
1 rows affected.


[(900, 2517, '35%')]

💡💡INSIGHTS <br>
• 35% of products sold by the company have unit prices higher than the average for their respective categories.

## LIMITATION OF THE DATASET
- Delivery date could not be used in the analysis due to a large number of null values.